### PR TITLE
Updating System Info Section

### LINKED ISSUE #2733: Default permissions on logs directory
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [x] Reporting an issue of an unmodified OSPOS installation
- [x] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [x] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [x] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [x] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [x] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [x] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 3.3.2
- OSPOS git commit hash is: 
- PHP version is: 
- MySQL or MariaDB version is: 
- OS and version is: 
- WebServer is: 
- Selected language is: 
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is: 

### Bug
Correct me if I'm wrong, but the default permissions given to the /application/logs/ directory is 755 or 775.  For security, shouldn't this be 750 or 770? I'm having a difficult time thinking of a scenario in which it's safe for world to have read access to the logs.

### PR DESCRIPTION
**Pull request by WebShells**
Updating System Info to check file permissions:
1st Case:
If all files permissions are set correctly no warnings will be displayed.
2nd Case:
If permissions weren't set correctly / But no security risks, a warning will be displayed
3rd Case:
If permissions were set correctly but causing security issues > 750, a warning will be displayed.

Clipboard function now fixed to work on Mozilla and other browsers. Text format now displayed in original format.
 closes #2733

### CODE REVIEW

**Comment by WebShells** [comment]
@objecttothis this should now be reflected to demo ?

**Comment by WebShells** [comment]
**Demo Currently Reflecting Updated Pull Request, results as follows:**

OSPOS Installation Info: 3.3.2 - 2dc06a

Language Code: en-US

Extensions & Modules:
» GD: Enabled ✓
» BC Math: Enabled ✓
» INTL: Enabled ✓
» OpenSSL: Enabled ✓
» MBString: Enabled ✓
» Curl: Enabled ✓

User Configuration:
.Browser: Chrome
.Server Software: Apache/2.4.38 (Debian)
.PHP Version: 7.3.17
.DB Version: 5.5.5-10.1.21-MariaDB-1~jessie
.Server Port: 80
.OS: Linux 87f1530e5f33 4.15.0-91-generic #92-Ubuntu SMP Fri Feb 28 11:09:48 UTC 2020 x86_64

File Permissions:
» [application/logs:] - 0775 | Writable ✓
» [public/uploads:] - 0755 | Writable ✓
» [public/uploads/item_pics:] - 0775 | Writable ✓
» [import_customers.csv:] - 0664 | Not Writable ✗

There are problems with file permissions please fix and reload this page.

Security Vulnerability Warning
Permissions higher than 750 leaves this software at risk.

=> [application/logs:] is writable, but the permissions are higher than 750.
=> [public/uploads:] is writable, but the permissions are higher than 750.
=> [public/uploads/item_pics:] is writable, but the permissions are higher than 750.

@opensourcepos/org

**Comment by jekkos** [comment]
> @objecttothis this should now be reflected to demo ?
> This is what travis says:
> Skipping a deployment with the bintray provider because the current build is a pull request.

Travis only pushes to bintray if the branch is on the opensourcepos repository.

**Comment by WebShells** [comment]
> > @objecttothis this should now be reflected to demo ?
> > This is what travis says:
> > Skipping a deployment with the bintray provider because the current build is a pull request.
> 
> Travis only pushes to bintray if the branch is on the opensourcepos repository.

Yes @jekkos noticed that, now it's fine... Created a new branch and created a new PR... All reflecting to the demo server as well...

File: application/views/configs/system_info.php [hunk 1]
```diff
@@ -12,94 +12,132 @@
...
 	</div>
-	</div>
 </div>
 <div align="center">
-		<a class="copy" data-clipboard-action="copy" data-clipboard-target="#issuetemplate">Copy Info</a> | <a href="https://github.com/opensourcepos/opensourcepos/issues/new" target="_blank"> Report An issue </a>		
-		<script>
-			var clipboard = new ClipboardJS('.copy');
+<a class="copy" data-clipboard-action="copy" data-clipboard-target="#issuetemplate">Copy Info</a> | <a href="https://github.com/opensourcepos/opensourcepos/issues/new" target="_blank"> Report An issue </a>		
+<script>
+	var clipboard = new ClipboardJS('.copy');
```
Comment (daN4cat) [inline application/views/configs/system_info.php:133]: Where is this Clipboard.js added to ospos?

File: application/views/configs/system_info.php [hunk 2]
```diff
@@ -12,94 +12,132 @@
...
 						}
-						clearstatcache();
-					?><br>
-					&#187; [public/uploads/item_pics:]	
-					<?php 
-						if (is_writable($images)) {
-							echo ' -  ' . substr(sprintf("%o",fileperms($images)),-4) . ' |	 ' . '<font color="green"> Writable &#x2713 </font>';
-						} else {
-							echo ' -  ' . substr(sprintf("%o",fileperms($images)),-4) . ' |	 ' . '<font color="red"> Not Writable &#x2717 </font>';
+						if(substr(decoct(fileperms($logs)), -4) >= 751
```
Comment (objecttothis) [inline application/views/configs/system_info.php:102]: After looking at the way the check displays on the screen in the demo, here are my thoughts:
- Items with writeable folders display a greet "writeable" and a green check, even if they have permissions vulnerabilities.

I recommend changing the functionality so that next to writeable or not writeable you place a red colored "vulnerable" or green "secure" and the red x appears if either one or the other is a problem.  This way, it is clear exactly which ones the warning below is referring to.  This would then treat the list at the top as the report and the list below as the summary to correct the top.  It works OK as is, so this doesn't have to be done, but would look simpler.

Something like (with appropriate colors):
» [application/logs:] - 0775 | Writable | Vulnerable ✗
» [public/uploads:] - 0750 | Writable | Secure ✓
» [public/uploads/item_pics:] - 0775 | Writable | Vulnerable ✗
» [import_customers.csv:] - 0664 | Not Writable ✗ 

There are problems with file permissions please fix and reload this page.

Security Vulnerability Warning
Permissions higher than 750 leaves this software at risk.

=> [application/logs:] is writable, but the permissions are higher than 750.
=> [public/uploads:] is writable, but the permissions are higher than 750.
=> [public/uploads/item_pics:] is writable, but the permissions are higher than 750.

- Small thing, but you should use the same » symbol at the bottom list as you do up top.
- Perhaps your summary at the bottom would have the non-writeable items in there as well?  blah blah is secure but not writeable.  Make the directory writeable by PHP.

**Comment by objecttothis** [comment]
> » [import_customers.csv:] - 0664 | Not Writable ✗

Does import_customers.csv need to be writeable?  I think it just needs to be readable by PHP, but perhaps I'm wrong.  We should still check the file to make sure it doesn't have security vulnerabilities... for that matter, if it were easy (not sure if it is), we could check all files recursively.  The only problem is that files in some directories need to be world readable and other should not be world readable.

File: application/views/configs/system_info.php [hunk 3]
```diff
@@ -12,94 +12,132 @@
...
-					<?php
-						if (is_writable($uploads)) {
-							echo ' -  ' . substr(sprintf("%o",fileperms($uploads)),-4) . ' |  ' . '<font color="green">	 Writable &#x2713 </font>';
-						} else {
-							echo ' -  ' . substr(sprintf("%o",fileperms($uploads)),-4) . ' |  ' . '<font color="red"> Not Writable &#x2717 </font>';
+				.Server Software: <?php echo $_SERVER['SERVER_SOFTWARE']; ?><br>
+				.PHP Version: <?php echo PHP_VERSION; ?><br>
+				.DB Version: <?php echo mysqli_get_server_info($this->db->conn_id); ?><br>
+				.Server Port: <?php echo $_SERVER['SERVER_PORT']; ?><br>
+				.OS: <?php echo php_uname();?><br><br>
```
Comment (objecttothis) [inline application/views/configs/system_info.php:46]: Is there any way to get a shorter version of the server OS version?  My string returns:

> .OS: FreeBSD [REDACTED] 11.3-RELEASE-p7 FreeBSD 11.3-RELEASE-p7 #0 r325575+ca0f1a6ba25(HEAD): Mon Apr 6 20:57:38 UTC 2020 [REDACTED]@tnbuild02.tn.ixsystems.com:/freenas-releng/freenas/_BE/objs/freenas-releng/freenas/_BE/os/sys/FreeNAS.amd64 amd64

When a better return would be "FreeBSD 11.3-RELEASE-p7"
Returning the domain name and the user that it's running as is dangerous because if people paste that information into an issue, it could leave them more vulnerable to attack.  Albeit, not likely successful, but perhaps most likely is a DOS or DDOS attack.

https://www.php.net/manual/en/function.php-uname.php suggests that perhaps instead you use the PHP_OS constant instead (https://www.php.net/manual/en/reserved.constants.php).  Try that and see if it gives you better results.

Comment (objecttothis) [reply application/views/configs/system_info.php:46]: Nevermind, use this instead
`.OS: <?php echo php_uname('s') .' '. php_uname('r');?><br><br>`
This will give you just the OS and version.  using PHP_OS only gives you the OS that PHP was built on which can be the same or it can be quite different.

File: application/views/configs/system_info.php (hunk 1 above)
Comment (WebShells) [reply application/views/configs/system_info.php:133]: @daN4cat under public/dist folder... Shall i add it through bower ? and get it minified with the build ?

File: application/views/configs/system_info.php (hunk 2 above)
Comment (WebShells) [reply application/views/configs/system_info.php:102]: @objecttothis am using a different symbol in order to differentiate or separate them... I can keep it the same no problem... Let me recheck your comments and will let you know if i need any clarification... Thanks

File: application/views/configs/system_info.php (hunk 1 above)
Comment (daN4cat) [reply application/views/configs/system_info.php:133]: I don't see it part of this PR. Not sure the code is completed....

Comment (WebShells) [reply application/views/configs/system_info.php:133]: @daN4cat i added it directly to master... In the previous submission ( copy to clipboard feature )

Comment (daN4cat) [reply application/views/configs/system_info.php:133]: Ok. I missed that. I need to review that commit.

Comment (WebShells) [reply application/views/configs/system_info.php:133]: @daN4cat am updating the PR to get up with @objecttothis comments... If you would like to wait until am done, or add your comments so i can do them all together...

File: application/views/configs/system_info.php (hunk 2 above)
Comment (WebShells) [reply application/views/configs/system_info.php:102]: Perhaps your summary at the bottom would have the non-writeable items in there as well? blah blah is secure but not writeable. Make the directory writeable by PHP.

With this won't we be repeating ourselves ? The results so far are as follows: 

OSPOS Installation Info: 3.3.1 - c786d4

Language Code: en-US

Extensions & Modules:
» GD: Enabled ✓
» BC Math: Enabled ✓
» INTL: Enabled ✓
» OpenSSL: Enabled ✓
» MBString: Enabled ✓
» Curl: Enabled ✓

User Configuration:
.Browser: Chrome
.Server Software: Apache/2.4.23 (Win64) PHP/7.0.10
.PHP Version: 7.0.10
.DB Version: 5.7.14
.Server Port: 80
.OS: Windows NT 6.1

File Permissions:
» [application/logs:] - 0777 | Writable ✓ | Vulnerable ✗
» [public/uploads:] - 0777 | Writable ✓ | Vulnerable ✗
» [public/uploads/item_pics:] - 0777 | Writable ✓ | Vulnerable ✗
» [import_customers.csv:] - 0666 | Readable ✓ | Security Check Passed ✓

There are problems with file permissions please fix and reload this page.

Security Vulnerability Warning
Permissions higher than 750 leaves this software at risk.

=> [application/logs:] is writable, but the permissions are higher than 750.
=> [public/uploads:] is writable, but the permissions are higher than 750.
=> [public/uploads/item_pics:] is writable, but the permissions are higher than 750.

Comment (WebShells) [reply application/views/configs/system_info.php:102]: This is an issue template @objecttothis, should be copied back to github... The shorter the better... I think so, is the above enough ? most of the comments were resolved..

**Comment by WebShells** [comment]
@jekkos the build failed with the below logs, can you check please ? something related to test tokens 
ospos_test | 1) Barcode_lib_test::test_barcode_weight_first
ospos_test | Error: Call to undefined method Barcode_lib::parse_barcode_fields()
ospos_test | 
ospos_test | /app/application/tests/libraries/Barcode_lib_test.php:20
ospos_test | 
ospos_test | 2) Barcode_lib_test::test_barcode_weight_last
ospos_test | Error: Call to undefined method Barcode_lib::parse_barcode_fields()
ospos_test | 
ospos_test | /app/application/tests/libraries/Barcode_lib_test.php:33
ospos_test | 
ospos_test | --
ospos_test | 
ospos_test | There were 3 failures:
ospos_test | 
ospos_test | 1) Token_lib_test::test_token_parser
ospos_test | Failed asserting that 'abc' matches expected null.
ospos_test | 
ospos_test | /app/application/tests/libraries/Token_lib_test.php:24
ospos_test | 
ospos_test | 2) Token_lib_test::test_barcode_item_number_and_quantity
ospos_test | Failed asserting that 123.456 matches expected 1.
ospos_test | 
ospos_test | /app/application/tests/libraries/Token_lib_test.php:38
ospos_test | 
ospos_test | 3) Token_lib_test::test_barcode_quantity_and_item_number
ospos_test | Failed asserting that 0.1 matches expected null.
ospos_test | 
ospos_test | /app/application/tests/libraries/Token_lib_test.php:52
ospos_test | 
ospos_test | ERRORS!
ospos_test | Tests: 10, Assertions: 10, Errors: 2, Failures: 3.
ospos_test exited with code 2
Stopping mysql      ... 
The command "docker-compose -f docker-compose.test.yml up --abort-on-container-exit" exited with 2.
Done. Your build exited with 1.

File: application/views/configs/system_info.php (hunk 2 above)
Comment (objecttothis) [reply application/views/configs/system_info.php:126]: > 
> 
> @objecttothis am using a different symbol in order to differentiate or separate them... I can keep it the same no problem... Let me recheck your comments and will let you know if i need any clarification... Thanks

@WebShells All the changes look great.  I still think you should replace "=>" with either "»" or another symbol.  I do understand your reasoning for wanting to differentiate/separate them, but the symbol you have looks like it belongs more in an ascii console than on a website.  If it were me, I would just use the same symbol because the different color and space difference separates them.  One thing you could do if you are wanting to separate them but link them is to use a superscript ¹, ², ³ below (instead of arrows and above to the right of the problem.  This gives it foot note connotation.  See below.

» [application/logs:] - 0777 | Writable ✓ | Vulnerable ✗ ¹
» [public/uploads:] - 0777 | Writable ✓ | Vulnerable ✗ ²
» [public/uploads/item_pics:] - 0777 | Writable ✓ | Vulnerable ✗ ³
» [import_customers.csv:] - 0666 | Readable ✓ | Security Check Passed ✓

There are problems with file permissions please fix and reload this page.

Security Vulnerability Warning
Permissions higher than 750 leaves this software at risk.

¹ [application/logs:] is writable, but the permissions are higher than 750.
² [public/uploads:] is writable, but the permissions are higher than 750.
³ [public/uploads/item_pics:] is writable, but the permissions are higher than 750.

File: application/views/configs/system_info.php [hunk 4]
```diff
@@ -12,82 +12,144 @@
...
+						echo "&#187; INTL: ", extension_loaded('intl') ? 'Enabled &#x2713' : 'MISSING &#x2717', '<br>';
+						echo "&#187; OpenSSL: ", extension_loaded('openssl') ? 'Enabled &#x2713' : 'MISSING &#x2717', '<br>';
+						echo "&#187; MBString: ", extension_loaded('mbstring') ? 'Enabled &#x2713' : 'MISSING &#x2717', '<br>';
+						echo "&#187; Curl: ", extension_loaded('curl') ? 'Enabled &#x2713' : 'MISSING &#x2717', '<br> <br>';		
+					?>
+				User Configuration:<br>
+				.Browser: <?php
+						function get_browser_name($user_agent)
+						{
+							if (strpos($user_agent, 'Opera') || strpos($user_agent, 'OPR/')) return 'Opera';
```
Comment (objecttothis) [inline application/views/configs/system_info.php:32]: https://www.geeksforgeeks.org/switch-vs-else/
This is probably better as a switch statement.  The way you have it is fine and it is more compact, but a switch could potentially be more readable.  This isn't bad and others may prefer it to a switch statement, so feel free to leave it as is, but I wanted to mention it.

File: application/views/configs/system_info.php [hunk 5]
```diff
@@ -12,82 +12,144 @@
...
-							echo ' -  ' . substr(sprintf("%o",fileperms($uploads)),-4) . ' |  ' . '<font color="green">	 Writable &#x2713 </font>';
-						} else {
-							echo ' -  ' . substr(sprintf("%o",fileperms($uploads)),-4) . ' |  ' . '<font color="red"> Not Writable &#x2717 </font>';
+				.Server Software: <?php echo $_SERVER['SERVER_SOFTWARE']; ?><br>
+				.PHP Version: <?php echo PHP_VERSION; ?><br>
+				.DB Version: <?php echo mysqli_get_server_info($this->db->conn_id); ?><br>
+				.Server Port: <?php echo $_SERVER['SERVER_PORT']; ?><br>
+				.OS: <?php echo php_uname('s') .' '. php_uname('r');?><br><br>
+				File Permissions:<br>
+						&#187; [application/logs:]
```
Comment (objecttothis) [inline application/views/configs/system_info.php:48]: It might just be github, but it looks like there is one too many tab indents here.

**Comment by jekkos** [comment]
@WebShells can you address the last open comments so we can proceed with the release? Thanks!

**Comment by WebShells** [comment]
I think we are ready to merge, finally we will have the bunch of tabs in configuration section merged together :x

File: application/language/th/config_lang.php [hunk 6]
```diff
@@ -276,8 +276,8 @@
 $lang["config_top"] = "ด้านบน";
 $lang["config_use_destination_based_tax"] = "เปิดใช้ Destination-Based Tax";
 $lang["config_website"] = "เว็บไซต์";
-$lang["config_work_order_enable"] = "";
-$lang["config_work_order_format"] = "";
+$lang["config_work_order_enable"] = "เปิดใช้ ใบสั่งงาน (WO)";
```
Comment (objecttothis) [inline application/language/th/config_lang.php:279]: Did you mean to include these changes in this PR?  It doesn't seem to be related to the System Info Section.

Comment (WebShells) [reply application/language/th/config_lang.php:279]: @objecttothis this was conflicting with the master, that's why i re-edited those lines.

Comment (WebShells) [reply application/language/th/config_lang.php:279]: So already submitted through weblate after i submitted the PR

Comment (objecttothis) [reply application/language/th/config_lang.php:279]: Makes sense.

**Comment by objecttothis** [comment]
> 
> 
> ```
> @objecttothis am using a different symbol in order to differentiate or separate them... I can keep it the same no problem... Let me recheck your comments and will let you know if i need any clarification... Thanks
> ```
> 
> @WebShells All the changes look great. I still think you should replace "=>" with either "»" or another symbol. I do understand your reasoning for wanting to differentiate/separate them, but the symbol you have looks like it belongs more in an ascii console than on a website. If it were me, I would just use the same symbol because the different color and space difference separates them. One thing you could do if you are wanting to separate them but link them is to use a superscript ¹, ², ³ below (instead of arrows and above to the right of the problem. This gives it foot note connotation. See below.
> 
> » [application/logs:] - 0777 | Writable ✓ | Vulnerable ✗ ¹
> » [public/uploads:] - 0777 | Writable ✓ | Vulnerable ✗ ²
> » [public/uploads/item_pics:] - 0777 | Writable ✓ | Vulnerable ✗ ³
> » [import_customers.csv:] - 0666 | Readable ✓ | Security Check Passed ✓
> 
> There are problems with file permissions please fix and reload this page.
> 
> Security Vulnerability Warning
> Permissions higher than 750 leaves this software at risk.
> 
> ¹ [application/logs:] is writable, but the permissions are higher than 750.
> ² [public/uploads:] is writable, but the permissions are higher than 750.
> ³ [public/uploads/item_pics:] is writable, but the permissions are higher than 750.

@WebShells did you see this suggestion I made regarding the formatting of the results?  I haven't downloaded the build so I didn't see if you made these changes or not.  No big deal if not, but I think it looks good.  Other than this, I have no other suggestions.

**Comment by WebShells** [comment]
> @WebShells did you see this suggestion I made regarding the formatting of the results? I haven't downloaded the build so I didn't see if you made these changes or not.

@objecttothis i applied same symbols/signs used in the previous sections.
